gh-140601: Add ResourceWarning to iterparse when not closed

osamakader opened the pull request:
When iterparse() opens a file by filename and is not explicitly closed, emit a ResourceWarning to alert developers of the resource leak.

This implements the TODO comment at line 1270 of ElementTree.py which has been requesting this feature since the close() method was added.

- Add _closed flag to IterParseIterator to track state
- Emit ResourceWarning in __del__ if not closed
- Add comprehensive test cases
- Update existing tests to properly close iterators

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
gh-NNNNNN: Summary of the changes made
```

Where: gh-NNNNNN refers to the GitHub issue number.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNNNN)
```

Where: [X.Y] is the branch name, for example: [3.13].

GH-NNNNNN refers to the PR number from `main`.

-->


<!-- gh-issue-number: gh-140601 -->
* Issue: gh-140601
<!-- /gh-issue-number -->

osamakader committed:
gh-140601: Add ResourceWarning to iterparse when not closed

When iterparse() opens a file by filename and is not explicitly closed,
emit a ResourceWarning to alert developers of the resource leak.

This implements the TODO comment at line 1270 of ElementTree.py which
has been requesting this feature since the close() method was added.

- Add _closed flag to IterParseIterator to track state
- Emit ResourceWarning in __del__ if not closed
- Add comprehensive test cases
- Update existing tests to properly close iterators

Signed-off-by: Osama Abdelkader <osama.abdelkader@gmail.com>

ashm-dev reviewed (commented):
.

ashm-dev commented on Lib/test/test_xml_etree.py:
```diff
@@ -692,28 +692,31 @@ def test_iterparse(self):
+        it = iterparse(SIMPLE_XMLFILE)
+        del it
+        import gc
+        gc.collect()  # Ensure previous iterator is cleaned up
```
Wouldn't it be better to use support.gc_collect() instead of directly calling gc.collect() was specifically added for use in tests to ensure consistent and deterministic garbage collection behavior across platforms and configurations.

serhiy-storchaka commented on Lib/xml/etree/ElementTree.py:
```diff
@@ -1261,18 +1261,35 @@ def iterator(source):
+                except:
+                    # Ignore errors during warning emission in __del__
+                    # This can happen during interpreter shutdown
+                    pass
```
This will unconditionally silence any bugs even if this happens not during interpreter shutdown. Look what other code emitting a ResourceWarning does.

serhiy-storchaka commented on Lib/xml/etree/ElementTree.py:
```diff
@@ -1261,18 +1261,35 @@ def iterator(source):
-            # (When the close() method will be supported in all maintained Python versions.)
+            if close_source and not getattr(self, '_closed', False):
+                try:
+                    warnings.warn(
```
Since `__del__` can be called during interpreter shutdown, it is better to not refer any globals. Look at `subprocess.Popen.__del__` for example.

serhiy-storchaka commented on Lib/xml/etree/ElementTree.py:
```diff
@@ -1261,18 +1261,35 @@ def iterator(source):
             if close_source:
                 source.close()
             gen.close()
+            self._closed = True
```
You can simply set `close_source = False`. But it should be declared `nonlocal`.

serhiy-storchaka commented on Lib/xml/etree/ElementTree.py:
```diff
@@ -1261,18 +1261,35 @@ def iterator(source):
         def __del__(self):
-            # TODO: Emit a ResourceWarning if it was not explicitly closed.
-            # (When the close() method will be supported in all maintained Python versions.)
+            if close_source and not getattr(self, '_closed', False):
```
Alternatively, you can test `gen.gi_frame is not None` or `not source.closed`. But setting `close_source = False` may be simpler and more reliable.

serhiy-storchaka reviewed (commented):
These tests were intentionally added to check that the resource is closed without explicit `close()`. Don't add `close()`. Just replace `check_no_resource_warning()` with `assertWarns(ResourceWarning)` and add `gc_collect()`. There are other tests that explicitly call `close()`.

osamakader committed:
Address review feedback from serhiy-storchaka

- Use nonlocal close_source instead of _closed flag
- Bind warnings.warn as default parameter in __del__
- Use assertWarns(ResourceWarning) instead of adding close() calls
- Use support.gc_collect() for consistent test behavior
- Remove broad exception handling

Signed-off-by: Osama Abdelkader <osama.abdelkader@gmail.com>

osamakader committed:
remove extra whitespace

Signed-off-by: Osama Abdelkader <osama.abdelkader@gmail.com>

osamakader committed:
remove unused import

Signed-off-by: Osama Abdelkader <osama.abdelkader@gmail.com>

serhiy-storchaka commented on Lib/test/test_xml_etree.py:
```diff
@@ -692,28 +692,33 @@ def test_iterparse(self):
+            next(it)
+        self.assertEqual(str(cm.exception),
+                'junk after document element: line 1, column 12')
+        with self.assertWarns(ResourceWarning):
```
Add `as cm` and check the warning message and filename. And the source if possible.

This is needed to ensure that the the resource was closed by the iterparse generator, not by the file destructor (which also emits a resource warning).

serhiy-storchaka commented on Lib/test/test_xml_etree.py:
```diff
@@ -692,28 +692,33 @@ def test_iterparse(self):
 
-        # Not exhausting the iterator still closes the resource (bpo-43292)
-        with warnings_helper.check_no_resource_warning(self):
+        # Deleting iterator without close() should emit ResourceWarning (bpo-43292)
```
Restore the old comment. The purpose of this test is to check that the resource is closed.

serhiy-storchaka commented on Lib/test/test_xml_etree.py:
```diff
@@ -725,6 +730,62 @@ def test_iterparse(self):
         with self.assertRaises(FileNotFoundError):
             iterparse("nonexistent")
 
+    def test_iterparse_resource_warning(self):
```
This all is already covered by the above tests, isn't?

osamakader replied:
yes, I just removed it.

osamakader committed:
Use % formatting instead of f-strings in __del__

F-strings can fail during cleanup. Use % formatting like
subprocess.Popen does for safer __del__ behavior.

Signed-off-by: Osama Abdelkader <osama.abdelkader@gmail.com>

osamakader committed:
Add assertions to verify ResourceWarning content

Check that warning message contains 'unclosed file' and the
filename to ensure it's from iterparse, not the file destructor.

Addresses review feedback from serhiy-storchaka.

Signed-off-by: Osama Abdelkader <osama.abdelkader@gmail.com>

osamakader committed:
Remove redundant test_iterparse_resource_warning function

The ResourceWarning tests are already covered in test_iterparse()
as noted by serhiy-storchaka in review.

Signed-off-by: Osama Abdelkader <osama.abdelkader@gmail.com>

osamakader committed:
Remove source=self from ResourceWarning

Removing source=self prevents unraisable exception in C extension.
The warning message still contains the filename via %r formatting.

Signed-off-by: Osama Abdelkader <osama.abdelkader@gmail.com>

serhiy-storchaka committed:
Add explicit close().

serhiy-storchaka committed:
Reorganize tests.

serhiy-storchaka committed:
Exhausted or failed iterators are auto-closed.

osamakader committed:
Add versionchanged directive for ResourceWarning

Document that iterparse now emits ResourceWarning in Python 3.15
when not explicitly closed.

Signed-off-by: Osama Abdelkader <osama.abdelkader@gmail.com>

serhiy-storchaka committed:
Merge branch 'main' into issue-140601-iterparse-warning

serhiy-storchaka committed:
Always emit a warning. Add tests for exhausted not closed iterator.

cmaloney commented on Doc/library/xml.etree.elementtree.rst:
```diff
@@ -656,6 +656,10 @@ Functions
    .. versionchanged:: 3.13
       Added the :meth:`!close` method.
 
+   .. versionchanged:: 3.15
```
```suggestion
   .. versionchanged:: next
```

cmaloney commented on Lib/test/test_xml_etree.py:
```diff
@@ -686,44 +693,96 @@ def test_iterparse(self):
+            list(it)
+            it.close()
+            del it, elem
+            gc_collect()
```
Should there also be a test here for when `iterparse` is passed a file object / result of an `open` to validate behavior that it leaves open files it dosen't own on `__del__`?

cmaloney commented on Lib/xml/etree/ElementTree.py:
```diff
@@ -1256,21 +1256,25 @@ def iterator(source):
         finally:
             if close_source:
-                source.close()
+                1#source.close()
```
Unintended change?

```suggestion
                source.close()
```

serhiy-storchaka replied:
Oh, it was for debugging. But since the tests are passed, I wonder how can we test that this line is present.

ashm-dev commented on Lib/xml/etree/ElementTree.py:
```diff
@@ -1256,21 +1256,25 @@ def iterator(source):
+        def __del__(self, _warn=warnings.warn):
             if close_source:
-                source.close()
+                try:
+                    _warn(f"unclosed iterparse iterator {source.name!r}", ResourceWarning, stacklevel=2)
+                finally:
+                    source.close()
```
Maybe this is better?
```suggestion
                name = getattr(source, 'name', None)
                if name:
                    _warn("unclosed iterparse iterator %r" % (name,), 
                          ResourceWarning, stacklevel=2)
                source.close()
```

serhiy-storchaka replied:
`source.name` always exist and not empty. This code is only executed when `iterparse()` opened a file by name.

BTW, if the `iterparse()` argument was file descriptor 0, your variant would not work.

ashm-dev replied:
You're right - I missed that `close_source=True` guarantees `source.name` exists, and `if name:` breaks on fd 0. Thanks for catching that! 

The original code is correct. (Though % formatting might be slightly safer than f-string in `__del__`, but not a blocker.)

ashm-dev commented on Lib/xml/etree/ElementTree.py:
```diff
@@ -1256,21 +1256,25 @@ def iterator(source):
+            nonlocal close_source
             if close_source:
                 source.close()
+                close_source = False
```
Wouldn't this option be better here?
```
try:
  if close_source:
    close_source = False
    source.close()
finally:
  gen.close()
```

serhiy-storchaka replied:
I wonder. It seems that only exception can be raised here is a KeyboardInterrupt. The difference between calling and not calling `gen.close()` is whether you can continue iterating after `iterparse().close()` (cached elements can still be yielded). But if `iterparse().close()` failed, you do not have any guarantees. Anyway, the file is closed first, so there will not be leaks. I think there will be no practical difference, so we can keep the simplest code.

ashm-dev commented on Lib/test/test_xml_etree.py:
```diff
@@ -686,44 +693,96 @@ def test_iterparse(self):
 
-        # Not exhausting the iterator still closes the resource (bpo-43292)
-        with warnings_helper.check_no_resource_warning(self):
+    def test_iterparse_not_close(self):
```
Wouldn't it be better to split one huge test into several smaller ones, since this test has already become too big?

serhiy-storchaka replied:
Maybe. There are 4 different cases here.

serhiy-storchaka replied:
There are other big tests here, so it is worth to refactor them before merging this PR. See #141499.

ashm-dev replied:
I agree

serhiy-storchaka committed:
Apply suggestions from code review

Co-authored-by: Cody Maloney <cmaloney@users.noreply.github.com>

serhiy-storchaka committed:
Refactor tests.

serhiy-storchaka committed:
Merge branch 'main' into issue-140601-iterparse-warning
